nav2_collision_monitor/src/source_base.cpp

```diff
@@ -0,0 +1,192 @@
+    RCLCPP_INFO(node->get_logger(), "Destroying SourceBase");
+  }
+
+  tf_buffer_.reset();
+}
+
+bool SourceBase::getParameters()
+{
+  auto node = node_.lock();
+  if (!node) {
+    throw std::runtime_error{"Failed to lock node"};
+  }
+
+  try {
+    nav2_util::declare_parameter_if_not_declared(
+      node, source_name_ + ".topic",
+      rclcpp::ParameterValue("scan"));  // Set deafult topic for laser scanner
+    source_topic_ = node->get_parameter(source_name_ + ".topic").as_string();
+  } catch (const std::exception & ex) {
+    RCLCPP_ERROR(
+      node->get_logger(), "Error while getting basic source parameters: %s", ex.what());
+    return false;
+  }
+
+  return true;
+}
+
+SourceType SourceBase::getSourceType()
+{
+  return source_type_;
+}
+
+std::string SourceBase::getSourceTypeStr()
+{
+  if (source_type_ == SOURCE_BASE) {
+    return "base";
+  } else if (source_type_ == SCAN) {
+    return "scan";
+  } else {  // source_type_ == POINTCLOUD
+    return "pointcloud";
+  }
+}
+
+bool SourceBase::getTransform(
+  const std::string to_frame,
+  const std::string from_frame,
+  tf2::Transform & tf2_transform)
+{
+  auto node = node_.lock();
+  if (!node) {
+    throw std::runtime_error{"Failed to lock node"};
+  }
+
+  if (!tf_buffer_) {
+    RCLCPP_ERROR(node->get_logger(), "tf_buffer_ is not initialized");
+    return false;
+  }
+
+  geometry_msgs::msg::TransformStamped transform;
+  tf2_transform.setIdentity();  // initialize by identical transform
+
+  if (from_frame == to_frame) {
+    // tf2_transform is already identical. Do nothing.
+    return true;
+  }
+
+  try {
+    const tf2::Duration tr_tol = tf2::durationFromSec(transform_tolerance_);
+    // Obtaining the transform to get data from one to another frame
+    transform = tf_buffer_->lookupTransform(
+      to_frame, from_frame, tf2::TimePointZero, tr_tol);
+  } catch (tf2::TransformException & e) {
+    RCLCPP_ERROR(
+      node->get_logger(),
+      "Failed to get \"%s\"->\"%s\" frame transform: %s",
+      source_frame_id_.c_str(), base_frame_id_.c_str(), e.what());
+    return false;
+  }
+
+  // Convert TransformStamped to TF2 transform
+  tf2::fromMsg(transform.transform, tf2_transform);
+  return true;
+}
+
+void SourceBase::getData(
+  std::vector<Point> & data, const rclcpp::Time & curr_time, const Velocity & velocity)
+{
+  std::lock_guard<mutex_t> lock(data_mutex_);
+
+  fixData(curr_time, velocity);
+  data.insert(data.end(), data_fixed_.data.begin(), data_fixed_.data.end());
+}
+
+void SourceBase::fixData(const rclcpp::Time & curr_time, const Velocity & velocity)
+{
+  auto node = node_.lock();
+  if (!node) {
+    throw std::runtime_error{"Failed to lock node"};
+  }
+
+  // Start opearting with data
+  std::lock_guard<mutex_t> lock(data_mutex_);
+  data_fixed_.data.clear();
+  data_fixed_.time = data_.time;
+
+  // Time shift between current time and time of latest data received.
+  // Used in movement correction of data
+  double dt;
+  try {
```

Comment:
Remove, this should never be a problem if clocks are set right, if they're not set right, there are many other issues

Reply:
This is strange, but I get this exception almost every time when re-run again `collision_monitor_node.launch.py` with continuing working `tb3_simulation_launch.py` simulation.

Reply:
Then the times aren't being set with the right clocks. Times and durations set as class variables will default to the wrong clock type if you don't specifically tell it on instantiation what to use https://github.com/ros-planning/navigation2/pull/2964/files#diff-42be06e95e3505a3b3006557312f35fe5102793df72baffa1e0fe6cb3415dd34R34

Reply:
This works for me. Thank you!

Reply:
Done for `velocity_stamp_` and `prev_time_`: https://github.com/ros-planning/navigation2/pull/2982/files#diff-e1c76d88179c14a1cf399e94bebd4f5d17dedca39e09b3bbdc810aa11a705543R31-R33. For `SourceBase` we do not need it (sure we could, but this will lead to again locking the node weak_ptr in constructors), since `data_` is [being initialized](https://github.com/ros-planning/navigation2/pull/2982/files#diff-a8ab77e548af62fe331c646d4df405f23654b0a652d6d63718a4c0c1f6ee1ab7R30) with `nullptr` and then [being checked](https://github.com/ros-planning/navigation2/pull/2982/files#diff-f467e09777b3dd7252b1f24d976c2049a8be9b317ac1fb391b80eb7829470f58R65) before any time stamp usage.